### ISSUE TITLE
Permission error on caches

### ISSUE THREAD

**Issue by Eijebong** [taskcluster/taskcluster#7404]
Since 75.0.0, I'm getting frequent permission errors when using caches.

```
[cache 2024-12-01T18:31:58.737Z] cache /builds/worker/checkouts exists; requirements: gid=1000 uid=1000 version=1
Traceback (most recent call last):
  File "/usr/local/bin/run-task", line 1359, in <module>
    sys.exit(main(sys.argv[1:]))
             ^^^^^^^^^^^^^^^^^^
  File "/usr/local/bin/run-task", line 1182, in main
    purge = configure_cache_posix(
            ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/bin/run-task", line 503, in configure_cache_posix
    write_audit_entry(audit_path, b"used")
  File "/usr/local/bin/run-task", line 329, in write_audit_entry
    with open(path, "ab") as fh:
         ^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/builds/worker/checkouts/.cachelog'
```

An example task: https://taskcluster.bananium.fr/tasks/HfpTR10xQWqJTuDVJq-PWg/runs/0

**Comment by Eijebong** [comment]
After reapplying #7149, I looked at it a bit more.

After the first run, the cache looks like this:

```
root@taskclusterium:/# ls -lah caches/F-q9yWvuSlGlepo27PG_3A/
total 20K
drwx------ 3   1018   1018 4.0K Dec 20 15:49 .
drwx------ 3 root   root   4.0K Dec 20 15:49 ..
-rw-r--r-- 1 165536 165536  106 Dec 20 15:49 .cachelog
-r--r--r-- 1 165536 165536   27 Dec 20 15:49 .cacherequires
drwxr-xr-x 7   1018   1018 4.0K Dec 20 15:49 vcs
```

Compared to some prod cache:

```
[root@tc-worker-medium-1 caches]# ls -lah DT1mWrz8SfW1JDwwM3Q-Tg/
total 20K
drwx------  3 root root 4.0K Dec  9 20:15 .
drwx------ 24 root root 4.0K Dec 14 19:52 ..
-rw-r--r--  1 1747 1747  216 Dec 10 08:22 .cachelog
-r--r--r--  1 1747 1747   21 Dec  9 20:15 .cacherequires
drwxr-xr-x  7 1747 1747 4.0K Dec 10 08:22 vcs
```

I wonder if maybe you're calling chown in a way that ignores dotfiles? I don't understand why it'd only break on docker though.

**Comment by bhearsum** [comment]
It looks like it just runs `chown -R`: https://github.com/taskcluster/taskcluster/pull/7149/files#diff-10d9fdf24fa3d9a913864112876b2360369fd4dfec4db5c104a97e2e826b7844R166

**Comment by Eijebong** [comment]
Alright, after scratching my head at https://github.com/taskcluster/taskcluster/issues/7427 for a while, I think I figured this one out too.


For the following tests I'm running commands with the following users:

- On the host, eijebong (1000:1000)
- In the container, root (0:0)
- In the container, user1 (1000:1000)
- In the container, user2 (1001:1001)
- 
I am mounting the directory `./caches` onto `/home/caches`. The directory is owned by `1000:1000` on the host with `0755` permissions.

I omitted unnecessary lines from commands output to keep it "short".

### Docker

Let's start with docker's behavior regarding volumes because it's the easiest.

Container started with `docker run -v ./caches:/home/caches -it debian:latest`

```
root@c97109eb3479:/# touch /home/caches/withroot
root@c97109eb3479:/# sudo -u user1 touch /home/caches/withuser1
root@c97109eb3479:/# sudo -u user2 touch /home/caches/withuser2
touch: cannot touch '/home/caches/withuser2': Permission denied
```

#### Volume state in the container
```
root@c97109eb3479:/# ls -lnah /home/
drwxr-xr-x 2 1000 1000 4.0K Dec 23 16:13 caches
root@c97109eb3479:/# ls -lnah /home/caches/
drwxr-xr-x 2 1000 1000 4.0K Dec 23 16:13 .
-rw-r--r-- 1    0    0    0 Dec 23 16:12 withroot
-rw-r--r-- 1 1000 1000    0 Dec 23 16:13 withuser1
```

#### Volume state on the host

```
[eijebong@plutonium] ~/cachetest  >>> ls -lnah caches
drwxr-xr-x 2 1000 1000 4.0K Dec 23 17:13 .
-rw-r--r-- 1    0    0    0 Dec 23 17:12 withroot
-rw-r--r-- 1 1000 1000    0 Dec 23 17:13 withuser1
```

#### So what's happening?

Docker doesn't do any mapping with UIDs, any file written by a UID in the container will be written with the same UID outside of the container. That means that tasks running with `run-task` where the `worker` user would write files as UID 1000 in the cache. And is reading the files back with UID 1000.

The linked PR "works" here because it's essentially a no-op for caches created with docker as it now tries changing ownership of files owned by a user with a UID far superior to 1000 (since the UID of the task user increases on every task creation).

### Podman without uidmaps

Let's talk podman now. First let's check what podman does if you run it just like docker without any extra argument

Container started with `podman run -v ./caches:/home/caches -it debian:latest`

```
root@7211237dcb51:/# touch /home/caches/withroot
root@7211237dcb51:/# sudo -u user1 touch /home/caches/withuser1
touch: cannot touch '/home/caches/withuser1': Permission denied
root@7211237dcb51:/# sudo -u user2 touch /home/caches/withuser2
touch: cannot touch '/home/caches/withuser2': Permission denied
```

#### Cache state in the container

```
root@7211237dcb51:/# ls -lanh /home/caches/
drwxr-xr-x 2 0 0 4.0K Dec 23 16:30 .
-rw-r--r-- 1 0 0    0 Dec 23 16:30 withroot
```

#### Cache state on the host

```
[eijebong@plutonium] ~/cachetest  >>> ls -lnah caches
drwxr-xr-x 2 1000 1000 4.0K Dec 23 17:30 .
-rw-r--r-- 1 1000 1000    0 Dec 23 17:30 withroot
```

#### What's happening?

We can see here that podman mapped the directory's owner from 1000 to 0 back and forth. Unfortunately this isn't very helpful, because we couldn't see what happened to normal users in that case. If I chmod `777` the cache directory on the host, then I get this:

```
root@8c6af2ca1163:/# ls -lnah /home/caches/
drwxrwxrwx 2    0    0 4.0K Dec 23 16:32 .
-rw-r--r-- 1    0    0    0 Dec 23 16:30 withroot
-rw-r--r-- 1 1000 1000    0 Dec 23 16:32 withuser1
```

And 

```
[eijebong@plutonium] ~/cachetest  >>> ls -lnah caches
drwxrwxrwx 2   1000   1000 4.0K Dec 23 17:32 .
-rw-r--r-- 1   1000   1000    0 Dec 23 17:30 withroot
-rw-r--r-- 1 100999 100999    0 Dec 23 17:32 withuser1
```

So by default, podman maps the current UID to 0 and then maps 1000 in the container to 100999 on the host. If we try to run podman as another user, using the same command, we get this in the container:

```
root@68e068afcfc4:/# ls -lnah /home/caches/
drwxrwxrwx 2 65534 65534 4.0K Dec 23 16:32 .
-rw-r--r-- 1 65534 65534    0 Dec 23 16:30 withroot
-rw-r--r-- 1 65534 65534    0 Dec 23 16:32 withuser1
```

That's why we need to map things ourselves with uidmap/gidmap, we don't want the cache to only be writable by root.

### Podman with id mapping

Container started with `podman run --uidmap 1000:0:1 --uidmap 0:1:1000 --uidmap 1001:1001:64536 --gidmap 1000:0:1 --gidmap 0:1:1000 --gidmap 1001:1001:64536 -v ./caches:/home/caches -it debian:latest` to match what d2g does.

```
root@903b2483090c:/# touch /home/caches/withroot
root@903b2483090c:/# sudo -u user1 touch /home/caches/withuser1
root@903b2483090c:/# sudo -u user2 touch /home/caches/withuser2
touch: cannot touch '/home/caches/withuser2': Permission denied
```

#### Cache state in the container

```
root@903b2483090c:/# ls -lanh /home/caches/
drwxr-xr-x 2 1000 1000 4.0K Dec 23 16:42 .
-rw-r--r-- 1    0    0    0 Dec 23 16:41 withroot
-rw-r--r-- 1 1000 1000    0 Dec 23 16:42 withuser1
```

#### Cache state on the host

```
[eijebong@plutonium] ~/cachetest  >>> ls -lnah caches
drwxr-xr-x 2   1000   1000 4.0K Dec 23 17:42 .
-rw-r--r-- 1 100000 100000    0 Dec 23 17:41 withroot
-rw-r--r-- 1   1000   1000    0 Dec 23 17:42 withuser1
```

#### What's happening

Well here we can see that the opposite happened from the default mapping. Instead the UID 1000 outside of the container got mapped to the UID 1000 inside. But root got mapped from 0 to 100000. Now what happens with this patch is half correct. When the cache gets given to another user, generic-worker will change the ownership of anything owned by UID `1000` to the new task user. It will however not touch the UID `100000`.

Let's quickly emulate that and try with another user.

```
[eijebong@plutonium] ~/cachetest  >>> sudo chown eijemoz:eijemoz caches/withuser1
[eijebong@plutonium] ~/cachetest  >>> sudo chown eijemoz:eijemoz caches
```

```
[eijebong@plutonium] ~/cachetest  >>> sudo -u eijemoz podman run --uidmap 1000:0:1 --uidmap 0:1:1000 --uidmap 1001:1001:64536 --gidmap 1000:0:1 --gidmap 0:1:1000 --gidmap 1001:1001:64536 -v ./caches:/home/caches -it debian:latest

root@a71664a3c153:/# ls -lnah /home/caches/
drwxr-xr-x 2  1000  1000 4.0K Dec 23 16:42 .
-rw-r--r-- 1 65534 65534    0 Dec 23 16:41 withroot
-rw-r--r-- 1  1000  1000    0 Dec 23 16:42 withuser1
```

We can see that it properly followed through for the files that were originally owned by UID `1000` but not for the ones owned by UID `0`.  The mapping is incorrect because my `eijebong (1000)` has the following subuid mapping: `eijebong:100000:65536` whereas my `eijemoz` user has the following: `eijemoz:231072:65536`. From that we can a "fix", by mapping the entire range of subuids as well as the original uid:

```
[eijebong@plutonium] ~/cachetest  >>> sudo chown eijemoz:eijemoz caches/withuser1
[eijebong@plutonium] ~/cachetest  >>> sudo chown eijemoz:eijemoz caches
[eijebong@plutonium] ~/cachetest  >>> sudo chown 231072:231072 caches/withroot
```

```
[eijebong@plutonium] ~/cachetest  >>> sudo -u eijemoz podman run --uidmap 1000:0:1 --uidmap 0:1:1000 --uidmap 1001:1001:64536 --gidmap 1000:0:1 --gidmap 0:1:1000 --gidmap 1001:1001:64536 -v ./caches:/home/caches -it debian:latest
root@cc5648c97172:/# ls -lnah /home/caches/
total 8.0K
drwxr-xr-x 2 1000 1000 4.0K Dec 23 16:42 .
-rw-r--r-- 1    0    0    0 Dec 23 16:41 withroot
-rw-r--r-- 1 1000 1000    0 Dec 23 16:42 withuser1
```

### TL;DR

The patch that got reverted "fixes" caches on docker because it essentially turns the cache ownership change into a no-op. Docker doesn't do any UID mapping so files gets written to the host as UID 0 and 1000, maybe 1001 in some cases when the task user's UID goes up all the time. 

It breaks podman caches because suddenly not all files get given to the new task user. Instead, only files that were owned by the UID 1000 user in the container get chowned. Anything owned by anyone else, gets mapped onto the subuid range on the host and their permissions stay as is.

A solution would be to do the mapping ourselves by looking at every file's UID, checking if it's in the previous owner's subuid range, then mapping it to the new user range.

**Comment by Eijebong** [comment]
A way better fix for this would be to always use the same subuid range for all task users.

At the moment they switch back and forth because every user addition is created 64k ahead of the previous one but it tries to fill holes.

```
root@5fce108af983:/# useradd user1
root@5fce108af983:/# useradd user2
root@5fce108af983:/# cat /etc/subuid
user1:100000:65536
user2:165536:65536
root@5fce108af983:/# userdel user1
root@5fce108af983:/# useradd user3
root@5fce108af983:/# cat /etc/subuid
user2:165536:65536
user3:100000:65536
```

Now if we create all task users with a subuid range like `user3:100000:65536`, then the patch actually has the intended effect for podman.

I tried locally by changing both my users to the same subuid/subgid range:

```
[eijebong@plutonium] ~/cachetest  >>> sudo -u eijemoz podman run --uidmap 1000:0:1 --uidmap 0:1:1000 --uidmap 1001:1001:64536 --gidmap 1000:0:1 --gidmap 0:1:1000 --gidmap 1001:1001:64536 -v ./caches:/home/caches -it debian:latest

root@5fce108af983:/# ls -lanh /home/caches/
total 8.0K
drwxr-xr-x 2  1000  1000 4.0K Dec 23 17:08 .
drwxr-xr-x 1     0     0 4.0K Dec 23 17:12 ..
-rw-r--r-- 1     0     0    0 Dec 23 17:08 withroot
-rw-r--r-- 1  1000  1000    0 Dec 23 17:08 withuser1
```

```
[eijebong@plutonium] ~/cachetest  >>> sudo chown eijebong:eijebong caches
[eijebong@plutonium] ~/cachetest  >>> sudo chown eijebong:eijebong caches/withuser1
[eijebong@plutonium] ~/cachetest  >>> podman run --uidmap 1000:0:1 --uidmap 0:1:1000 --uidmap 1001:1001:64536 --gidmap 1000:0:1 --gidmap 0:1:1000 --gidmap 1001:1001:64536 -v ./caches:/home/caches -it debian:latest
root@1331422a13ae:/# ls -lnah /home/caches/
total 8.0K
drwxr-xr-x 2 1000 1000 4.0K Dec 23 17:08 .
drwxr-xr-x 3    0    0 4.0K Dec 23 17:15 ..
-rw-r--r-- 1    0    0    0 Dec 23 17:08 withroot
-rw-r--r-- 1 1000 1000    0 Dec 23 17:08 withuser1
```

**Comment by petemoore** [comment]
Thanks for the deep dive Bastien!
